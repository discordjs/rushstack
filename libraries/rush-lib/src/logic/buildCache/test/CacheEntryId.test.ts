// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

jest.mock('node:process', () => {
  return {
    ...jest.requireActual('node:process'),
    platform: 'dummyplatform',
    arch: 'dummyarch'
  };
});

import { CacheEntryId, type GetCacheEntryIdFunction } from '../CacheEntryId';

describe(CacheEntryId.name, () => {
  describe('Valid pattern names', () => {
    describe.each([
      { projectName: 'project+name', note: 'without a scope' },
      { projectName: '@scope/project+name', note: 'with a scope' }
    ])('For a project name $note', ({ projectName }) =>
      it.each([
        undefined,
        '[hash]',
        '[projectName]_[hash]',
        '[phaseName:normalize]_[hash]',
        '[phaseName:trimPrefix]_[hash]',
        '[projectName:normalize]_[hash]',
        '[projectName:normalize]_[phaseName:normalize]_[hash]',
        '[projectName:normalize]_[phaseName:normalize]_[hash]_[os]_[arch]',
        '[projectName:normalize]_[phaseName:trimPrefix]_[hash]',
        'prefix/[projectName:normalize]_[hash]',
        'prefix/[projectName:normalize]_[phaseName:normalize]_[hash]',
        'prefix/[projectName:normalize]_[phaseName:trimPrefix]_[hash]',
        'prefix/[projectName]_[hash]',
        'prefix/[projectName]_[phaseName:normalize]_[hash]',
        'prefix/[projectName]_[phaseName:trimPrefix]_[hash]'
      ])('Handles pattern %s', (pattern) => {
        const getCacheEntryId: GetCacheEntryIdFunction = CacheEntryId.parsePattern(pattern);
        expect(
          getCacheEntryId({
            projectName,
            projectStateHash: '09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3',
            phaseName: '_phase:compile'
          })
        ).toMatchSnapshot();
      })
    );
  });

  describe('Invalid pattern names', () => {
    it.each([
      'x',
      '[invalidTag]',
      'unstartedTag]',
      '[incompleteTag',
      '[hash:badAttribute]',
      '[hash:badAttribute:attr2]',
      '[projectName:badAttribute]',
      '[projectName:]',
      '[phaseName]',
      '[phaseName:]',
      '[phaseName:badAttribute]',
      '[:attr1]',
      '[projectName:attr1:attr2]',
      '/[hash]',
      '[os:attr]',
      '[arch:attr]',
      '~'
    ])('Throws an exception for an invalid pattern (%s)', (pattern) => {
      expect(() => CacheEntryId.parsePattern(pattern)).toThrowErrorMatchingSnapshot();
    });
  });
});
