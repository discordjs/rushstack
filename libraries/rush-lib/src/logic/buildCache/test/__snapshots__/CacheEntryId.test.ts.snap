// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern (/[hash]) 1`] = `"Cache entry name patterns may not start with a slash."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern ([:attr1]) 1`] = `"Unexpected token name \\"\\"."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern ([arch:attr]) 1`] = `"An attribute isn't supported for the \\"arch\\" token."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern ([hash:badAttribute:attr2]) 1`] = `"An attribute isn't supported for the \\"hash\\" token."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern ([hash:badAttribute]) 1`] = `"An attribute isn't supported for the \\"hash\\" token."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern ([incompleteTag) 1`] = `"Unclosed token in cache entry name pattern."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern ([invalidTag]) 1`] = `"Unexpected token name \\"invalidTag\\"."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern ([os:attr]) 1`] = `"An attribute isn't supported for the \\"os\\" token."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern ([phaseName:]) 1`] = `"Unexpected attribute \\"\\" for the \\"phaseName\\" token."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern ([phaseName:badAttribute]) 1`] = `"Unexpected attribute \\"badAttribute\\" for the \\"phaseName\\" token."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern ([phaseName]) 1`] = `"Either the \\"normalize\\" or the \\"trimPrefix\\" attribute is required for the \\"phaseName\\" token."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern ([projectName:]) 1`] = `"Unexpected attribute \\"\\" for the \\"projectName\\" token."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern ([projectName:attr1:attr2]) 1`] = `"Unexpected attribute \\"attr1:attr2\\" for the \\"projectName\\" token."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern ([projectName:badAttribute]) 1`] = `"Unexpected attribute \\"badAttribute\\" for the \\"projectName\\" token."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern (~) 1`] = `"Cache entry name pattern contains an invalid character. Only alphanumeric characters, slashes, underscores, and hyphens are allowed."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern (unstartedTag]) 1`] = `"Unexpected \\"]\\" character in cache entry name pattern."`;

exports[`CacheEntryId Invalid pattern names Throws an exception for an invalid pattern (x) 1`] = `"Cache entry name pattern is missing a [hash] token."`;

exports[`CacheEntryId Valid pattern names For a project name with a scope Handles pattern [hash] 1`] = `"09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name with a scope Handles pattern [phaseName:normalize]_[hash] 1`] = `"_phase_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name with a scope Handles pattern [phaseName:trimPrefix]_[hash] 1`] = `"compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name with a scope Handles pattern [projectName:normalize]_[hash] 1`] = `"scope+project++name_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name with a scope Handles pattern [projectName:normalize]_[phaseName:normalize]_[hash] 1`] = `"scope+project++name__phase_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name with a scope Handles pattern [projectName:normalize]_[phaseName:normalize]_[hash]_[os]_[arch] 1`] = `"scope+project++name__phase_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3_dummyplatform_dummyarch"`;

exports[`CacheEntryId Valid pattern names For a project name with a scope Handles pattern [projectName:normalize]_[phaseName:trimPrefix]_[hash] 1`] = `"scope+project++name_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name with a scope Handles pattern [projectName]_[hash] 1`] = `"@scope/project+name_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name with a scope Handles pattern prefix/[projectName:normalize]_[hash] 1`] = `"prefix/scope+project++name_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name with a scope Handles pattern prefix/[projectName:normalize]_[phaseName:normalize]_[hash] 1`] = `"prefix/scope+project++name__phase_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name with a scope Handles pattern prefix/[projectName:normalize]_[phaseName:trimPrefix]_[hash] 1`] = `"prefix/scope+project++name_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name with a scope Handles pattern prefix/[projectName]_[hash] 1`] = `"prefix/@scope/project+name_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name with a scope Handles pattern prefix/[projectName]_[phaseName:normalize]_[hash] 1`] = `"prefix/@scope/project+name__phase_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name with a scope Handles pattern prefix/[projectName]_[phaseName:trimPrefix]_[hash] 1`] = `"prefix/@scope/project+name_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name with a scope Handles pattern undefined 1`] = `"09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name without a scope Handles pattern [hash] 1`] = `"09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name without a scope Handles pattern [phaseName:normalize]_[hash] 1`] = `"_phase_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name without a scope Handles pattern [phaseName:trimPrefix]_[hash] 1`] = `"compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name without a scope Handles pattern [projectName:normalize]_[hash] 1`] = `"project++name_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name without a scope Handles pattern [projectName:normalize]_[phaseName:normalize]_[hash] 1`] = `"project++name__phase_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name without a scope Handles pattern [projectName:normalize]_[phaseName:normalize]_[hash]_[os]_[arch] 1`] = `"project++name__phase_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3_dummyplatform_dummyarch"`;

exports[`CacheEntryId Valid pattern names For a project name without a scope Handles pattern [projectName:normalize]_[phaseName:trimPrefix]_[hash] 1`] = `"project++name_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name without a scope Handles pattern [projectName]_[hash] 1`] = `"project+name_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name without a scope Handles pattern prefix/[projectName:normalize]_[hash] 1`] = `"prefix/project++name_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name without a scope Handles pattern prefix/[projectName:normalize]_[phaseName:normalize]_[hash] 1`] = `"prefix/project++name__phase_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name without a scope Handles pattern prefix/[projectName:normalize]_[phaseName:trimPrefix]_[hash] 1`] = `"prefix/project++name_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name without a scope Handles pattern prefix/[projectName]_[hash] 1`] = `"prefix/project+name_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name without a scope Handles pattern prefix/[projectName]_[phaseName:normalize]_[hash] 1`] = `"prefix/project+name__phase_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name without a scope Handles pattern prefix/[projectName]_[phaseName:trimPrefix]_[hash] 1`] = `"prefix/project+name_compile_09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;

exports[`CacheEntryId Valid pattern names For a project name without a scope Handles pattern undefined 1`] = `"09d1ecee6d5f888fa6c35ca804b5dac7c3735ce3"`;
